language/move-stdlib/sources/string.move

```diff
@@ -0,0 +1,99 @@
+/// The `string` module defines the `String` type which represents UTF8 encoded strings.
+module std::string {
+    use std::vector;
+    use std::option::{Self, Option};
+
+    /// An invalid UTF8 encoding.
+    const EINVALID_UTF8: u64 = 1;
+
+    /// Index out of range.
+    const EINVALID_INDEX: u64 = 2;
+
+    /// A `String` holds an immutable sequence of bytes which is guaranteed to be in utf8 format.
+    struct String has copy, drop, store {
+        bytes: vector<u8>,
```

Comment:
Totally fine with this for now, but is this BCS compatible with BCS utf8 strings?

One benefit of doing a native string is that BCS will verify that the string is utf8, which is really nice to have for transaction/`entry` function arguments.

Reply:
Totally understand the benefit of a native string type. But is is also very costly to implement :-/ You mentioned before when we discussed impl of u256 that it is difficult to add a new type to the system, and sure it is. 

The deserialization problem is the only one I see for strings which a native type could do better. Memory representation should be good (and actually being able to "move " `vector<u8>` into a string is awesome for perf).

But we need to solve the problem of invariants on transaction parameters anyway, and String is only one special case of those.  I think we need some kind of allow list on structs for transactions, and then custom validators for them.

Reply:
> But we need to solve the problem of invariants on transaction parameters anyway, and String is only one special case of those. I think we need some kind of allow list on structs for transactions, and then custom validators for them.

It is indeed something we need to solve. And we should. 

My hope was that if the BCS serialization format would be the same for this struct  and the BCS strings, that we could drop in replace this string struct with a native type over time.